actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb

```diff
@@ -0,0 +1,77 @@
+require 'active_support/core_ext/object/try'
+
+module ActionView
+  module CollectionCaching # :nodoc:
+    extend ActiveSupport::Concern
+
+    included do
+      # Fallback cache store if Action View is used without Rails.
+      # Otherwise overriden in Railtie to use Rails.cache.
+      mattr_accessor(:collection_cache) { ActiveSupport::Cache::MemoryStore.new }
+    end
+
+    private
+
+    def cache_collection_render
+      return yield unless cache_collection?
+
+      keyed_collection = collection_by_cache_keys
+      partial_cache = collection_cache.read_multi(*keyed_collection.keys)
+
+      @collection = keyed_collection.reject { |key, _| partial_cache.key?(key) }.values
+      rendered_partials = @collection.any? ? yield : []
+
+      fetch_or_cache_partial(partial_cache, order_by: keyed_collection.each_key) do
+        rendered_partials.shift
+      end
+    end
+
+    def cache_collection?
+      if ActionController::Base.perform_caching
+        @options[:cache].present? || automatic_cache_eligible?
+      end
+    end
+
+    def automatic_cache_eligible?
+      if single_template_render? && !callable_cache_key?
+        @template.source =~ /\A<% cache\(?\s*(\w+\.?)/ && $1 == inferred_cache_name.to_s
+      end
+    end
+
+    def single_template_render?
+      @template # Template is only set when a collection renders one template.
+    end
+
+    def inferred_cache_name
+      @options[:as] || @template.virtual_path.split('/').last.sub('_', '')
+    end
+
+    def callable_cache_key?
+      @options[:cache].respond_to?(:call)
+    end
+
+    def collection_by_cache_keys
+      seed = callable_cache_key? ? @options[:cache] : ->(i) { i }
+
+      Hash[ @collection.map { |item| [expanded_cache_key(seed.call(item)), item] } ]
+    end
+
+    def expanded_cache_key(key)
+      caching_controller.fragment_cache_key(@view.cache_fragment_name(key)).tap do |k|
+        return k.dup if k.frozen? # #read_multi & #write may require mutability, Dalli 2.6.0.
+      end
+    end
+
+    def caching_controller
+      defined?(@_controller) ? @_controller : (@_caching_controller ||= ActionController::Base.new)
+    end
+
+    def fetch_or_cache_partial(cache, order_by:)
+      cache_options = @options[:cache_options] || @locals[:cache_options] || {}
+
+      order_by.map do |key|
+        cache[key] || yield.tap { |item| collection_cache.write(key, item, cache_options) }
```

Comment:
Right on. Does it make sense to check for `cache_options` in both `@options` or `@locals`? It was in the original `multi_fetch_fragments` implementation and I don't have enough experience with this to judge it with.